src/core/or/connection_edge.c

```diff
@@ -943,14 +936,49 @@ connection_edge_finished_connecting(edge_connection_t *edge_conn)
     int connected_payload_len =
       connected_cell_format_payload(connected_payload, &conn->addr,
                                     edge_conn->address_ttl);
-    if (connected_payload_len < 0)
+    if (connected_payload_len < 0) {
+      if (zero_payload_cleanup) {
```

Comment:
I'm not sure I like this new parameter whose only purpose is to run some code that is present in `connection_exit_connect()` but absent in `connection_edge_finished_connecting()`. It seems to make the supposed de-duplication effort more complicated than it has to be. Could `connection_exit_connect()` maybe check for a `-1` return from `connection_exit_set_state_open()` instead? It seems to me that there's no other place that `connection_exit_set_state_open()` can return `-1`. (A return of `-1` from `connection_edge_send_command()` maps to `0`.)